here-naksha-lib-ext-manager/src/main/java/com/here/naksha/lib/extmanager/ExtensionCache.java

```diff
@@ -83,15 +85,38 @@ protected void buildExtensionCache(ExtensionConfig extensionConfig) {
       publishIntoCache(result, extensionConfig);
     });
 
-    // Removing existing extension which has been removed from the configuration
-    List<String> extIds =
-        extensionConfig.getExtensions().stream().map(Extension::getId).toList();
+    // Removing existing extension which has been removed from the configuration.
+    List<String> extIds = extensionConfig.getExtensions().stream()
+        .map(extension -> extension.getEnv() + ":" + extension.getId())
+        .toList();
 
     for (String key : loaderCache.keySet()) {
       if (!extIds.contains(key)) {
-        loaderCache.remove(key);
-        PluginCache.removeExtensionCache(key);
-        logger.info("Extension {} removed from cache.", key);
+        KVPair<Extension, ClassLoader> kvPair = loaderCache.get(key);
+        Object instanceObj = kvPair.getValue();
+        Extension extension = kvPair.getKey();
+        final String extensionIdWthEnv = extension.getEnv() + ":" + extension.getId();
+        if (!isNullOrEmpty(extension.getInitClassName())) {
+          if (instanceObj instanceof IExtensionInit initInstance) {
+            try {
+              initInstance.close();
+              instance = initInstance;
+              logger.info("Extension {} closed successfully.", extensionIdWthEnv);
+            } catch (Exception e) {
+              logger.error("Failed to close extension {}", extensionIdWthEnv, e);
+            }
+          } else {
+            logger.error("Instance is not of type IExtensionInit for extension {}", extensionIdWthEnv);
+          }
+        }
+        synchronized (this) {
```

Comment:
I don't think we need synchronized here... as this function is not expected to be called by multiple threads in parallel

Reply:
Yes agree. Corrected it.